### ISSUE TITLE
Memory leak in EVP_DigestInit_ex

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#17149]
I think I found a memory leak in `EVP_DigestInit_ex`

https://github.com/openssl/openssl/blob/6f87463b62f9b2849510d74ff0fd6a62955ea947/crypto/evp/digest.c#L219-L222

We invoke `OPENSSL_clear_free` that works well for blocks of memory that don't contain pointers. But if the opaque context contains pointers to a memory, we get memory leak here. To avoid it, we should call the `ctx->digest->cleanup` function.

This bug was found because OpenVPN uses `EVP_DigestInit_ex` to reset their mac context, and GOST OMAC-based digests allocate CMAC_CTX objects for their purposes.

**Comment by beldmit** [comment]
Reproducer for 1.1.1:

```
#include <stdio.h>
#include <string.h>
#include <openssl/evp.h>

typedef struct {
        void *pdata;
} tmp_dgst_ctx;

static int _init(EVP_MD_CTX *ctx)
{
        tmp_dgst_ctx *p = EVP_MD_CTX_md_data(ctx);
        p->pdata = OPENSSL_malloc(32);
}

static int _cleanup(EVP_MD_CTX *ctx)
{
        tmp_dgst_ctx *p = EVP_MD_CTX_md_data(ctx);
        OPENSSL_free(p->pdata);
        p->pdata = NULL;
}
int main(int argc, char *argv[])
{
    EVP_MD_CTX *mdctx;
    const EVP_MD *md;
    EVP_MD *tmp;
    char mess1[] = "Test Message\n";
    char mess2[] = "Hello World\n";
    unsigned char md_value[EVP_MAX_MD_SIZE];
    unsigned int md_len, i;

    tmp = EVP_MD_meth_new(NID_id_GostR3411_94, NID_undef);
    EVP_MD_meth_set_init(tmp, _init);
    EVP_MD_meth_set_cleanup(tmp, _cleanup);
    EVP_MD_meth_set_app_datasize(tmp, sizeof(tmp_dgst_ctx)); 

    md = EVP_get_digestbyname("sha256");
    if (md == NULL) {
        exit(1);
    }

    mdctx = EVP_MD_CTX_new();
    EVP_DigestInit_ex(mdctx, tmp, NULL);
    EVP_DigestInit_ex(mdctx, md, NULL);
    EVP_DigestUpdate(mdctx, mess1, strlen(mess1));
    EVP_DigestUpdate(mdctx, mess2, strlen(mess2));
    EVP_DigestFinal_ex(mdctx, md_value, &md_len);
    EVP_MD_CTX_free(mdctx);

    printf("Digest is: ");
    for (i = 0; i < md_len; i++)
        printf("%02x", md_value[i]);
    printf("\n");

    EVP_MD_meth_free(tmp);

    exit(0);
}
```
Couldn't reproduce for 3.0, BTW - possibly because of legacy/non-legacy stuff

**Comment by t8m** [comment]
> Couldn't reproduce for 3.0, BTW - possibly because of legacy/non-legacy stuff

That's quite interesting. I'd expect this to be reproducible with 3.0 as well as invoking EVP_DigestInit_ex with md from EVP_MD_meth_new is a legacy path and should be equivalent to 1.1.1.

**Comment by beldmit** [comment]
I also was surprised

**Comment by beldmit** [comment]
Is it possible to elaborate a fix to the upcoming release? At a first glance, calling the `cleanup` should resolve it...

**Comment by mattcaswell** [comment]
> That's quite interesting. I'd expect this to be reproducible with 3.0 as well as invoking EVP_DigestInit_ex with md from EVP_MD_meth_new is a legacy path and should be equivalent to 1.1.1.

Unfortunately it seems that 3.0 is not detecting the MD from `EVP_MD_meth_new` as being legacy, so the first `EVP_DigestInit_ex` call goes down the provider path and subsequently fails. Since there isn't a return check on that call to `EVP_DigestInit_ex` that failure goes unnoticed in the reproducer.

**Comment by mattcaswell** [comment]
Draft fix in #17255

**Comment by beldmit** [comment]
Reopen as we need a fix for 1.1.1
